test/PaletteModeUtilTest.cc

```diff
@@ -0,0 +1,312 @@
+ *
+ * Expected result:
+ * Every pixels are closest to their centroid in all candidates
+ *
+ * Test coverage:
+ * Tests for K from PALETTE_MIN_SIZE to PALETTE_MAX_SIZE
+ */
+class KMeansTest : public ::testing::TestWithParam<int> {
+  protected:
+    KMeansTest() : rnd_(8, false), palette_rnd_(2, 64) {
+        k_ = GetParam();
+        data_ = new int[2 * MAX_PALETTE_SQUARE];
+    }
+
+    ~KMeansTest() {
+        if (data_) {
+            delete[] data_;
+            data_ = nullptr;
+        }
+    }
+
+    /** functions for 1d test */
+    int prepare_data(const int max_colors) {
+        memset(data_, 0, MAX_PALETTE_SQUARE * sizeof(int));
+        uint8_t *palette = new uint8_t[max_colors];
+        for (int i = 0; i < max_colors; i++)
+            palette[i] = rnd_.random();
+        uint8_t tmp[MAX_PALETTE_SQUARE] = {0};
+        for (size_t i = 0; i < MAX_PALETTE_SQUARE; i++)
+            data_[i] = tmp[i] = palette[rnd_.random() % max_colors];
+        delete[] palette;
+        int val_count[MAX_PALETTE_SQUARE] = {0};
+        return eb_av1_count_colors(tmp, 64, 64, 64, val_count);
+    }
+
+    void run_test(size_t times) {
+        uint8_t indeces[MAX_PALETTE_SQUARE] = {0};
+        for (size_t i = 0; i < times; i++) {
+            const int max_colors = palette_rnd_.random();
+            const int colors = prepare_data(max_colors);
+            int centroids[PALETTE_MAX_SIZE] = {0};
+            int k = AOMMIN(colors, k_);
+            av1_k_means_dim1(
+                data_, centroids, indeces, MAX_PALETTE_SQUARE, k, MaxItr);
+            check_output(centroids, k, data_, indeces, MAX_PALETTE_SQUARE);
+        }
+    }
+
+    void check_output(const int *centroids, const int k, const int *data,
+                      const uint8_t *indeces, const int n) {
+        for (int i = 0; i < n; i++) {
+            const int min_delta = abs(data[i] - centroids[indeces[i]]);
+            for (int j = 0; j < k; j++) {
+                const int delta = abs(data[i] - centroids[j]);
+                ASSERT_GE(delta, min_delta)
+                    << "index error at " << i << ", value is " << data[i]
+                    << ", distance to centroid( " << centroids[indeces[i]]
+                    << ") is greater than to " << centroids[j];
+            }
+        }
+    }
+
+    /** functions for 2d test */
+    int prepare_data_2d(const int max_colors) {
+        memset(data_, 0, 2 * MAX_PALETTE_SQUARE * sizeof(int));
+        uint16_t *palette = new uint16_t[max_colors];
+        for (int i = 0; i < max_colors; i++)
+            palette[i] = (rnd_.random() << 8) + rnd_.random();
+        vector<uint16_t> val_vec;
+        for (size_t i = 0; i < MAX_PALETTE_SQUARE; i++) {
+            uint16_t tmp = palette[rnd_.random() % max_colors];
+            data_[2 * i] = tmp >> 8;
+            data_[2 * i + 1] = tmp & 0xFF;
+            val_vec.push_back(tmp);
+        }
+        delete[] palette;
+        std::sort(val_vec.begin(), val_vec.end());
```

Comment:
Why not use count_color?

Reply:
the aim of test is to check the result of "count_color", it seems not good to involve the target function to help create the input data.

Reply:
and the count_color only support bit-depth 12-bit, here the bit-depth is 16-bit, count_color reports error